docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+
+In the current deployment model, control plane guarantees that a tenant is attached to one
+pageserver at a time, thereby ruling out split-brain conditions.
+However, there is always the risk of a control plane bug.
+
+Futher, lack of safety during split-brain conditions blocks two important features where occasional
+split-brain conditions are part of the design assumptions:
+* seamless tenant migration ([RFC PR](https://github.com/neondatabase/neon/pull/5029))
+* automatic pageserver instance failure handling (aka "failover") (RFC TBD)
+
+### Prior art
+
+- 020-pageserver-s3-coordination.md
+- 023-the-state-of-pageserver-tenant-relocation.md
+- 026-pageserver-s3-mvcc.md
+
+This RFC has broad similarities to the proposal to implement a MVCC scheme in
+S3 object names, but this RFC avoids a general purpose transaction scheme in
+favour of more specialized "generations" that work like a transaction ID that
+always has the same lifetime as a pageserver process and/or tenant attachment.
+
+## Requirements
+
+- Accommodate storage backends with no atomic or fencing capability (i.e. work within
+  S3's limitation that there are no atomics and clients can't be fenced)
+- Don't depend on any STONITH or node fencing in the compute layer (i.e. we will not
+  assume that we can reliably kill and EC2 instance and have it die)
+- Scoped per-tenant, not per-pageserver; for *seamless tenant migration*, we need
+  per-tenant granularity, and for *failover*, we likely want to spread the workload
+  of the failed pageserver instance to a number of peers, rather than monolithically
+  moving the entire workload to another machine.
+  We do not rule out the latter case, but should not constrain ourselves to it.
+
+## Design Tenets
+
+These are not requirements, but are ideas that guide the following design:
+
+- Avoid implementing another consensus system: we already have a strongly consistent
+  database in the control plane that can do atomic operations where needed, and we also
+  have a Paxos implementation in the safekeeper.
+- Avoiding locking in to specific models of how failover will work (e.g. do not assume that
+  all the tenants on a pageserver will fail over as a unit).
+- Avoid doing synchronization that scales with the number of tenants, unless absolutely
+  necessary.
+- Be strictly correct when it comes to data integrity. Occasional failures of availability
+  are tolerable, occasional data loss is not.
+
+## Non Goals
+
+The changes in this RFC intentionally isolate the design decision of how to define
+logical generations IDs and object storage format in a way that is somewhat flexible with
+respect to how actual orchestration of failover works.
+
+This RFC intentionally does not cover:
+
+- Failure detection
+- Orchestration of failover
+- Standby modes to keep data ready for fast migration
+- Intentional multi-writer operation on tenants (multi-writer scenarios are assumed to be transient split-brain situations).
+- Sharding.
```

Comment:
We do want most of those things though. And I think this proposal has been written with these things in mind. While this RFC doesn't explain how these would work, the assumption is that they *can* be implemented on top of this.

> Intentional multi-writer operation on tenants (multi-writer scenarios are assumed to be transient split-brain situations).

Hmm. There is a different writer postgres node on each timeline. So it's per-timeline, not per-tenant. Should we consider that somehow? We don't need to support multiple writer pageservers on a single timeline, but supporting a different writer for different timelines within the same tenant would make sense. If you create 1000 branches, and start loading data to each of them, one pageserver won't be able to keep up with the writes.

We don't need to solve that here, but let's make sure we don't paint ourselves in the corner.

Reply:
>We do want most of those things though. 

Of course, but the RFC is already really long :-)  I added a link here to Appendix B, where I discuss how this RFC intersects with other features.

> We don't need to support multiple writer pageservers on a single timeline, but supporting a different writer for different timelines within the same tenant would make sense. If you create 1000 branches, and start loading data to each of them, one pageserver won't be able to keep up with the writes.

If we did want to introduce the ability to attach timelines in the same tenant on different pageservers, I think that would work as an extension to this RFC: there is no requirement that the generation suffixes in S3 keys are really per-tenant, they could be issued per-timeline without changing anything at the storage layer.

My hunch is that key sharding will get us most of what we need for this kind of high bandwidth case though: given a choice between doing key sharding _and_ timeline distribution across nodes, vs. doing finer key-sharding so that even with many timelines the work per shard is manageable, I'd lean toward the latter.  But if we wanted to do it, then issuing timeline attachment generations instead of tenant attachment generations could happen pretty organically at the same time as changing the CP logic for how attachments work.